### PR TITLE
Add failing test: after hook is not run if test skipped in beforeEach

### PR DESCRIPTION
**Pull request by dasilvacontin**
Related to #2286.

### CODE REVIEW

**Comment by boneskull** [comment]
@dasilvacontin I think this fixes it, but I need a sanity check

**Comment by boneskull** [comment]
cc @mochajs/mocha or anyone else who wants to test this out

**Comment by boneskull** [comment]
as per discussion, I've updated my code to be more weird, but also more backwards-compatible.

cc @mochajs/mocha

**Comment by ScottFreeCode** [comment]
I'm wary of depending on the `run` interface when all we want to test is hooks, and of splitting the test out into a separate file so you have to look in those two separate places to figure it out, but the only alternative I can think of will throw from an after or after each hook in case of failure instead (which is arguably an abuse):
```js
describe('issue-2286: after doesn\'t execute if test was skipped in beforeEach', function () {
  var afterWasRun = false
  describe('suite with skipped test for meta test', function () {
    beforeEach(function () { this.skip() })
    after(function () { afterWasRun = true })
    it('should be pending', function () {})
  })
  after('meta test', function () {
    afterWasRun.should.be.ok();
  })
})
```

File: test/integration/fixtures/regression/issue-2286.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+describe('suite', function () {
+  beforeEach(function () { this.skip() })
+  after(function () { console.log('after in suite') })
+  it('test', function () {})
+})
```
Comment (ScottFreeCode) [inline test/integration/fixtures/regression/issue-2286.js]: How about some names that make this file's function obvious (both in viewing this file and if it were to end up run among the main tests in addition to the meta test calling it by `run`)? e.g.:
describe: 'testception for issue 2286 meta test'
beforeEach: 'skips tests'
after: 'should run'
it: 'skipped by beforeEach'

File: lib/runner.js [hunk 2]
```diff
@@ -302,7 +302,13 @@ Runner.prototype.hook = function(name, fn) {
       }
       if (err) {
         if (err instanceof Pending) {
-          suite.pending = true;
+          if (name === 'beforeEach' || name === 'afterEach') {
+            self.test.pending = true;
+          } else {
+            suite.tests.forEach(function(test) {
+              test.pending = true;
+            });
```
Comment (ScottFreeCode) [inline lib/runner.js]: Couldn't it use `suite.pending` if it's not `beforeEach` or `afterEach`? Or in the older versions does `after` run even if `before` calls this.skip()?

File: test/integration/fixtures/regression/issue-2286.js (hunk 1 above)
Comment (boneskull) [inline test/integration/fixtures/regression/issue-2286.js]: @ScottFreeCode I don't see this stuff as absolutely necessary given the filename, but if you wish to add to the branch, go for it

Comment (boneskull) [inline test/integration/fixtures/regression/issue-2286.js]: actually on second thought, I like your test above.  please use it instead

File: lib/runner.js (hunk 2 above)
Comment (boneskull) [inline lib/runner.js]: setting the suite to be pending, at this point, will cause the `after` hook to not get executed.

Comment (boneskull) [inline lib/runner.js]: (so, the answer is "no")

File: test/integration/fixtures/regression/issue-2286.js (hunk 1 above)
Comment (ScottFreeCode) [inline test/integration/fixtures/regression/issue-2286.js]: (to the first reply:) Yeah, it's pretty clear from the filename what issue this is testing, but it's not so obvious to me from looking just at this file that this is meant to be run by another test in another file that tests the output of this one (unless that's what all the stuff in this fixtures folder is -- I haven't looked at the files that were already there), so it looked at first as though this weren't actually testing it programmatically but only printing something to look at manually.

Comment (ScottFreeCode) [inline test/integration/fixtures/regression/issue-2286.js]: (to the second:) You mean the one that uses nested suites with code like this in the inner one and a test in the `after` of the outer one?

Comment (ScottFreeCode) [inline test/integration/fixtures/regression/issue-2286.js]: (Went ahead and committed both; feel free to rebase away either you don't want.)

File: test/integration/regression.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-var assert = require('assert');
+﻿var assert = require('assert');
 var fs     = require('fs');
```
Comment (ScottFreeCode) [inline test/integration/regression.js]: Did this... pick up a UTF-8 BOM or something? I will see if I can fix that when I get another free moment.

Comment (ScottFreeCode) [inline test/integration/regression.js]: Or not, if it's already merged into the history and isn't hurting anything...

**Pull request merged**